pkg/experiment/metastore/index/index.go

```diff
@@ -0,0 +1,462 @@
+// looking up blocks with FindBlock() or FindBlocksInRange().
+func NewIndex(store Store, logger log.Logger, cfg *Config) *Index {
+
+	// A fixed cache size gives us bounded memory footprint, however changes to the partition duration could reduce
+	// the cache effectiveness.
+	// TODO (aleks-p):
+	//  - resize the cache at runtime when the config changes
+	//  - consider auto-calculating the cache size to ensure we hold data for e.g., the last 24 hours
+	cache, _ := lru.New[PartitionKey, *indexPartition](cfg.PartitionCacheSize)
+
+	return &Index{
+		loadedPartitions: cache,
+		allPartitions:    make([]*PartitionMeta, 0),
+		store:            store,
+		logger:           logger,
+		Config:           cfg,
+	}
+}
+
+// LoadPartitions reads all partitions from the backing store and loads the recent ones in memory.
+func (i *Index) LoadPartitions() {
+	i.partitionMu.Lock()
+	defer i.partitionMu.Unlock()
+
+	i.allPartitions = make([]*PartitionMeta, 0)
+	for _, key := range i.store.ListPartitions() {
+		pMeta := i.loadPartitionMeta(key)
+		i.allPartitions = append(i.allPartitions, pMeta)
+	}
+
+	i.sortPartitions()
+}
+
+func (i *Index) loadPartitionMeta(key PartitionKey) *PartitionMeta {
+	t, dur, _ := key.Parse()
+	pMeta := &PartitionMeta{
+		Key:       key,
+		Ts:        t,
+		Duration:  dur,
+		Tenants:   make([]string, 0),
+		tenantMap: make(map[string]struct{}),
+	}
+	for _, s := range i.store.ListShards(key) {
+		for _, t := range i.store.ListTenants(key, s) {
+			pMeta.AddTenant(t)
+		}
+	}
+	return pMeta
+}
+
+// ForEachPartition executes the given function concurrently for each partition. It will be called for all partitions,
+// regardless if they are fully loaded in memory or not.
+func (i *Index) ForEachPartition(ctx context.Context, fn func(meta *PartitionMeta) error) error {
+	i.partitionMu.Lock()
+	defer i.partitionMu.Unlock()
+
+	g, ctx := errgroup.WithContext(ctx)
+	for _, meta := range i.allPartitions {
+		g.Go(func() error {
+			return fn(meta)
+		})
+	}
+	err := g.Wait()
+	if err != nil {
+		level.Error(i.logger).Log("msg", "error during partition iteration", "err", err)
+		return err
+	}
+	return nil
+}
+
+func (i *Index) getPartition(meta *PartitionMeta) *indexPartition {
+	p, ok := i.loadedPartitions.Get(meta.Key)
+	if !ok {
+		level.Info(i.logger).Log("msg", "loading partition", "key", meta.Key)
+		p = &indexPartition{
+			meta:     meta,
+			shards:   make(map[uint32]*indexShard),
+			loadedAt: time.Now(),
+		}
+		i.loadedPartitions.Add(meta.Key, p)
+		for _, s := range i.store.ListShards(meta.Key) {
+			sh := &indexShard{
+				tenants: make(map[string]*indexTenant),
+			}
+			p.shards[s] = sh
+
+			for _, t := range i.store.ListTenants(meta.Key, s) {
+				te := &indexTenant{
+					blocks: make(map[string]*metastorev1.BlockMeta),
+				}
+				for _, b := range i.store.ListBlocks(meta.Key, s, t) {
+					te.blocks[b.Id] = b
+				}
+				sh.tenants[t] = te
+			}
```

Comment:
I believe we should avoid loading all tenant for the partition (I just can't find a use case)

Reply:
This is now done, together with the cache key change.